diff --git a/multiupload_filefield_widget.module b/multiupload_filefield_widget.module
index f797cd9..2f81d31 100644
--- a/multiupload_filefield_widget.module
+++ b/multiupload_filefield_widget.module
@@ -33,10 +33,25 @@ function multiupload_filefield_widget_element_info() {
       'js' => array($mfw_path . '/mfw.js', $file_path . '/file.js'),
   ),
   );
+
+  if (module_exists('filefield_sources')) {
+    $types['mfw_managed_file']['#process'][] = 'filefield_sources_field_process';
+    $types['mfw_managed_file']['#pre_render'][] = 'filefield_sources_field_pre_render';
+    $types['mfw_managed_file']['#element_validate'][] = 'filefield_sources_field_validate';
+    $types['mfw_managed_file']['#file_value_callbacks'][] = 'filefield_sources_field_value';
+  }
+
   return $types;
 }
 
 /**
+ * Implements hook_filefield_sources_widgets().
+ */
+function multiupload_filefield_widget_filefield_sources_widgets() {
+  return array('file_mfw');
+}
+
+/**
  * Process function to expand the mfw_managed_file element type.
  *
  * Expands the file type to include Upload and Remove buttons, as well as
